db.cpp

```diff
@@ -175,22 +175,31 @@ void CAddrDb::GetIPs_(set<CNetAddr>& ips, int max, const bool* nets) {
     } else {
       id = *ourId.begin();
     }
-    if (id >= 0) {
+    if (id >= 0 && idToInfo[id].services & requestedFlags) {
       ips.insert(idToInfo[id].ip);
     }
     return;
   }
-  if (max > goodId.size() / 2)
-    max = goodId.size() / 2;
+  std::set<int> goodIdFiltered;
+  for (std::set<int>::const_iterator it = goodId.begin(); it != goodId.end(); it++) {
```

Comment:
Using a vector instead of a set would be more efficient, as we already know they're all different anyway.

The whole logic below that uses range/low/high can go away then, and instead pick `goodIdFiltered[rand() % goodIdFiltered.size()]`.